File: 프로그래머스/2/72412. 순위 검색/순위 검색.js
const CATEGORIES = {
    languages: [ "java", "python", "cpp" ],
    positions: [ "backend", "frontend" ],
    experienceLevels: [ "junior", "senior" ],
    favioriteFoods: [ "chicken", "pizza" ],
};


const fillQuery = (completeQueries, oldQuery, index) => {
    if (index === oldQuery.length) return completeQueries;
    if (oldQuery[index] === '-') {
        const categoryEntries = Object.entries(CATEGORIES);
        const [ _, categoryList] = categoryEntries[index]
        
        const newQueries = completeQueries.flatMap((completeQuery) => {
            return categoryList.map((option) => completeQuery.concat(option));
        });
        
        
        return fillQuery(newQueries, oldQuery, index + 1);
    }
    
    const newQueries = completeQueries.map((completeQuery) => completeQuery.concat(oldQuery[index]));
    return fillQuery(newQueries, oldQuery, index + 1);
}

function solution(info, queries) {
    const scoreboard = new Map();
    info.forEach((personString) => {
        const parts = personString.split(" ");
        const score = Number(parts.pop());
        const fields = parts; // [lang, pos, exp, food]

      // mask 0~15: 0b0000 ~ 0b1111
        for (let mask = 0; mask < 16; mask++) {
            const keyParts = [];
            for (let i = 0; i < 4; i++) {
                // 비트 i 가 켜져 있으면 원본, 꺼져 있으면 '-'
                keyParts.push((mask & (1 << i)) ? fields[i] : '-');
            }
            const key = keyParts.join(" ");
            if (!scoreboard.has(key)) scoreboard.set(key, []);
            scoreboard.get(key).push(score);
        }
    });
    
    scoreboard.forEach((value, key, map) => map.get(key).sort((a, b) => a - b));
    
    return queries.map((query) => {
        const [ language, and1, position, and2, experienceLevel, and3, favoriteFood, scoreString ] = query.split(" ");
        
        const key = [ language, position, experienceLevel, favoriteFood ].join(" ");
            const target = Number(scoreString);
            const list = scoreboard.get(key);

            if (!list || list.length === 0 || list[list.length - 1] < target) return 0;
            
            let start = 0;
            let end = list.length - 1;
            while (start < end) {
                const middle = Math.floor((start + end) / 2);
                if (target <= list[middle]) {
                    end = middle;
                } else { // list[middle] < target
                    start = middle + 1;
                }
            }
            const lowerbound = start;
            return list.length - lowerbound;
            
    });
}


// 트리 사용시: 100,000 * 24 * 4(분류 찾는데 사용) * log2 2000(탐색) => 약 10억번 연산으로 시간초과 충분히 가능
// 따라서, 그냥 매핑태이블을 사용해서 분류 찾는데 사용되는 시간을 줄여야 한다 => 100,000 * log 2000

File: 프로그래머스/2/72412. 순위 검색/순위 검색.kt
class Solution {
    fun solution(info: Array<String>, queries: Array<String>): IntArray {
        val scoreTable = info.flatMap { 
            val tokens = it.split(" ")
            val options = tokens.slice(0 until tokens.size - 1)
            val score = tokens.last().toInt()
            
            val pairs = (0..15).map { mask ->
                val key = options.mapIndexed { index, option ->
                    when (mask and (1 shl index)) {
                        0 -> '-'
                        else -> option
                    }
                }.joinToString(" ")
                key to score
            }
            pairs
        }.fold(mutableMapOf<String, MutableList<Int>>()) { acc, (key, score) ->
            if (!(key in acc)) {
                acc[key] = mutableListOf<Int>()
            }
            acc[key]!!.add(score)
            acc
        }
        scoreTable.values.map { mutableList ->
            mutableList.sort()
        }
        
        
        
        val result = queries.asSequence().map { query ->
            val tokens = query.split(" ").filter { it != "and" }
            val options = tokens.slice(0 until tokens.size - 1)
            val key = options.joinToString(" ")
            val score = tokens.last().toInt()
            
            val list = scoreTable[key] ?: emptyList<Int>()
            val count = if (list.isEmpty() || score > list.last()) {
                0
            } else {
                var start = 0
                var end = list.size - 1


                while (start < end) {
                    val middle = (start + end) / 2
                    // lowerbound = min(i | list[i] >= score)
                    if (score <= list[middle]) { // lowerbound <= list[middle]
                        end = middle
                    } else { // list[middle] < score, max(j | list[j] < score) < min(i | list[i] >= score) = lowerbound
                        start = middle + 1
                    }
                }
                val lowerbound = end
                list.size - lowerbound      
            }
            count
        }.toList()
        
        return result.toIntArray()
    }
}


// len(info) = nlen(query) = m
// mapping table에 담음 O(n), n <= 50_000
// 각 query별로 이진탐색해서 계산 => O(mlogn) * 24 